mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -96,14 +127,321 @@ impl EthCoin {
+        let swap_id = self.etomic_swap_id(payment_time_lock, args.maker_secret_hash);
+        let taker_status = self
+            .payment_status_v2(
+                taker_swap_v2_contract,
+                Token::FixedBytes(swap_id.clone()),
+                &TAKER_SWAP_V2,
+                EthPaymentType::TakerPayments,
+                3,
+            )
+            .await?;
+
+        let tx_from_rpc = self.transaction(TransactionId::Hash(args.funding_tx.tx_hash())).await?;
+        let tx_from_rpc = tx_from_rpc.as_ref().ok_or_else(|| {
+            ValidateSwapV2TxError::TxDoesNotExist(format!(
+                "Didn't find provided tx {:?} on ETH node",
+                args.funding_tx.tx_hash()
+            ))
+        })?;
+        validate_from_to_and_status(
+            tx_from_rpc,
+            taker_address,
+            taker_swap_v2_contract,
+            taker_status,
+            TakerPaymentStateV2::PaymentSent as u8,
+        )?;
+
+        let validation_args = {
+            let dex_fee = wei_from_big_decimal(&args.dex_fee.fee_amount().into(), self.decimals)?;
+            let payment_amount = wei_from_big_decimal(&(args.trading_amount + args.premium_amount), self.decimals)?;
+            TakerValidationArgs {
+                swap_id,
+                amount: payment_amount,
+                dex_fee,
+                receiver: self.my_addr().await,
+                taker_secret_hash: args.taker_secret_hash,
+                maker_secret_hash: args.maker_secret_hash,
+                funding_time_lock,
+                payment_time_lock,
+            }
+        };
+        match self.coin_type {
+            EthCoinType::Eth => {
+                let function = TAKER_SWAP_V2.function(ETH_TAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_eth_taker_payment_data(&decoded, &validation_args, function, tx_from_rpc.value)?;
+            },
+            EthCoinType::Erc20 { token_addr, .. } => {
+                let function = TAKER_SWAP_V2.function(ERC20_TAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_erc20_taker_payment_data(&decoded, &validation_args, function, token_addr)?;
+            },
+            EthCoinType::Nft { .. } => unreachable!(),
         }
+        Ok(())
+    }
+
+    pub(crate) async fn taker_payment_approve(
+        &self,
+        args: &GenTakerFundingSpendArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let (taker_swap_v2_contract, send_func, token_address) = self
+            .taker_swap_v2_details(ETH_TAKER_PAYMENT, ERC20_TAKER_PAYMENT)
+            .await?;
+        let decoded = try_tx_s!(decode_contract_call(send_func, args.funding_tx.unsigned().data()));
+        // Note: `PaymentSent` state was checked in `validate_taker_funding`. So we don't have to validate it in prepare function
```

Comment:
we should still do it tho 👀 if you can of course..

Reply:
technically we can do it, there is no prob, decoded data contains swap id, its all what we need to call the state from swap contract.
I understand your concern, but we will send unnecessary calls to the EVM if we add  the same checks to next stage.

Right now process looks like this:
 - Taker sends payment to smart contract (**send_taker_funding**): `Uninitialized`-> `PaymentSent`
 - Maker validates taker payment (**validate_taker_funding**): `PaymentSent` check
 - Taker approves payment (**sign_and_send_taker_funding_spend**): `PaymentSent` -> `TakerApproved`
 - Maker checks approve (**search_for_taker_funding_spend**): `TakerApproved` check
 - Maker spends taker payment (**sign_and_broadcast_taker_payment_spend**): `TakerApproved` -> `MakerSpent`

Adding additional checks will result in the following:
 - Taker sends payment to smart contract (**send_taker_funding**): `Uninitialized`-> `PaymentSent`
 - Maker validates taker payment (**validate_taker_funding**): `PaymentSent` check
 - Taker approves payment (**sign_and_send_taker_funding_spend** function): `PaymentSent` check, `PaymentSent` -> `TakerApproved`
 - Maker checks approve (**search_for_taker_funding_spend**): `TakerApproved` check
 - Maker spends taker payment (**sign_and_broadcast_taker_payment_spend**): `TakerApproved` check, `TakerApproved` -> `MakerSpent`

I also would like to remove payment state check in nft maker implementations, as we do the payment validation and then in next step we do the same check again. But now I see that if state machine didnt fail on prev stage, than the data should be correct.

Something the same I see in utxo swap implementation, expected outputs and preimages are checked in separate `validation_...` methods

I think both approaches have their own pros and cons, I would like to follow the first variant.
Lets clarify it with @shamardy

Reply:
@borngraced Ive added payment states re check https://github.com/KomodoPlatform/komodo-defi-framework/pull/2169/commits/a8594144ba0348665f327144db9d295e92eb7f9a
we still have to complete maker ops, so we have time to think should we leave it or revert